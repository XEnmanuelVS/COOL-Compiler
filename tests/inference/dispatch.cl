class Main {
    infer_int(): AUTO_TYPE {
        (new A).int()
    };
    infer_string(): AUTO_TYPE {
        (new A).string()
    };
};

class A {
    int(): AUTO_TYPE {
        new Int
    };
    string(): AUTO_TYPE {
        new String
    };
};
