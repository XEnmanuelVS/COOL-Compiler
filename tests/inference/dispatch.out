\__ProgramNode [<class> ... <class>]
  \__ClassDeclarationNode: class Main  { <feature> ... <feature> }
    \__FuncDeclarationNode: infer_int() : Int -> <body>
      \__CallNode: <obj>.int(<expr>, ..., <expr>)
        \__InstantiateNode: new A()

    \__FuncDeclarationNode: infer_string() : String -> <body>
      \__CallNode: <obj>.string(<expr>, ..., <expr>)
        \__InstantiateNode: new A()

  \__ClassDeclarationNode: class A  { <feature> ... <feature> }
    \__FuncDeclarationNode: int() : Int -> <body>
      \__InstantiateNode: new Int()
    \__FuncDeclarationNode: string() : String -> <body>
      \__InstantiateNode: new String()
